s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -65,6 +65,31 @@ private[alpakka] final class S3Stream(credentials: AWSCredentials,
     Source.fromFuture(request(s3Location, range).flatMap(entityForSuccess).map(_.dataBytes)).flatMapConcat(identity)
   }
 
+  def listBucket(bucket: String, prefix: Option[String] = None): Source[String, NotUsed] = {
+    import system.dispatcher
+
+    /**
+     *
+     * @param args The Option[String] is our continuation token and the Boolean is our holder
+     *             for wether this is the first time the method was called. When the token is None
+     *             and the first call Boolean is false, we know that we are done getting keys.
+     * @return
+     */
+    def listBucketCall(args: (Option[String], Boolean)): Future[Option[((Option[String], Boolean), Seq[String])]] =
```

Comment:
Ah, I see, you need to distinguish "Start", "Running", and "End" states. Maybe it would make sense to create a three state ADT instead? That would read a bit cleaner:

```scala
sealed trait State
case object Starting extends State
case class Running(continuationTokenL String) extends State
case object Finished extends State
```

WDYT?